src/librustc_mir/borrow_check/mod.rs

```diff
@@ -708,6 +702,66 @@ impl InitializationRequiringAction {
 }
 
 impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+
+    fn visit_terminator_drop(
+        &mut self,
+        loc: Location,
+        term: &Terminator<'tcx>,
+        flow_state: &Flows<'cx, 'gcx, 'tcx>,
+        drop_place: &Place<'tcx>,
+        span: Span,
+    ) {
+        let ty = drop_place.ty(self.mir, self.tcx).to_ty(self.tcx);
+        match ty.sty {
+            // When a struct is being dropped, we need to check whether it has a
+            // destructor, if it does, then we can call it, if it does not then we
+            // need to check the individual fields instead.
+            // See #47703.
+            ty::TyAdt(def, substs) if def.is_struct() && !def.has_dtor(self.tcx) => {
+                for (index, field) in def.all_fields().enumerate() {
+                    let field_ty = field.ty(self.tcx, substs);
+                    let proj = Projection {
+                        base: drop_place.clone(),
+                        elem: ProjectionElem::Field(Field::new(index), field_ty),
+                    };
+                    let place = Place::Projection(Box::new(proj));
+
+                    match field_ty.sty {
+                        // It may be the case that this issue occurs with a struct within a
+                        // struct, so we recurse to handle that.
+                        ty::TyAdt(def, _) if def.is_struct() && !def.has_dtor(self.tcx) => {
+                            self.visit_terminator_drop(
+                                loc,
+                                term,
+                                flow_state,
+                                &place,
+                                span,
+                            );
+                        },
+                        _ => {
+                            self.access_place(
+                                ContextKind::Drop.new(loc),
+                                (&place, span),
+                                (Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
```

Comment:
In particular, this `Shallow(None)` is wrong -- we should fall through to the `Deep` case below. 

Here is why: imagine this *is* a struct with a destructor. Then we will be running random user code, it could touch anything within!

Reply:
I wasn't sure about the `Shallow(None)`, but when using the place for just the field, we still get an error, so figured it must have been necessary:

```
DEBUG:<unknown>: MirBorrowckCtxt::process_terminator(bb0[4], Terminator { source_info: SourceInfo { span: src/test/run-pass/issue-47703.rs:25:50: 25:50, scope: scope[0] }, kind: drop(_1) -> [return: bb2, unwind: bb1] }): borrows in effect: [&mut (*(_1.0: &mut ())), &mut (*(_1.0: &mut ()))@active] borrows generated: [] inits: [_0, _1] uninits: [_2] move_out: [] ever_init: [mp1@src/test/run-pass/issue-47703.rs:25:16: 25:20 (Deep), mp0@src/test/run-pass/issue-47703.rs:25:36: 25:50 (Deep)]
DEBUG:<unknown>: check_access_permissions((_1.0: &mut ()), Write(StorageDeadOrDrop), Yes)
DEBUG:<unknown>: places_conflict((*(_1.0: &mut ())),(_1.0: &mut ()),Deep)
DEBUG:<unknown>: places_conflict: components [_1, (_1.0: &mut ()), (*(_1.0: &mut ()))] / [_1, (_1.0: &mut ())]
DEBUG:<unknown>: places_conflict: Some(_1) vs. Some(_1)
DEBUG:<unknown>: place_element_conflict: DISJOINT-OR-EQ-LOCAL
DEBUG:<unknown>: places_conflict: Some((_1.0: &mut ())) vs. Some((_1.0: &mut ()))
DEBUG:<unknown>: place_element_conflict: DISJOINT-OR-EQ-FIELD
DEBUG:<unknown>: places_conflict: Some((*(_1.0: &mut ()))) vs. None
DEBUG:<unknown>: places_conflict: None vs. None
DEBUG:<unknown>: places_conflict: full borrow, CONFLICT
```

Reply:
I'm able to change this so that the `access_place` ends up matching the `borrow_place`, ie. `(*(_1.0: &mut ()))` on both sides. This also does not fix the error.

```
DEBUG:<unknown>: MirBorrowckCtxt::process_terminator(bb0[4], Terminator { source_info: SourceInfo { span: src/test/run-pass/issue-47703.rs:25:50: 25:50, scope: scope[0] }, kind: drop(_1) -> [return: bb2, unwind: bb1] }): borrows in effect: [&mut (*(_1.0: &mut ())), &mut (*(_1.0: &mut ()))@active] borrows generated: [] inits: [_0, _1] uninits: [_2] move_out: [] ever_init: [mp1@src/test/run-pass/issue-47703.rs:25:16: 25:20 (Deep), mp0@src/test/run-pass/issue-47703.rs:25:36: 25:50 (Deep)]
DEBUG:<unknown>: check_access_permissions((*(_1.0: &mut ())), Write(StorageDeadOrDrop), Yes)
DEBUG:<unknown>: places_conflict((*(_1.0: &mut ())),(*(_1.0: &mut ())),Deep)
DEBUG:<unknown>: places_conflict: components [_1, (_1.0: &mut ()), (*(_1.0: &mut ()))] / [_1, (_1.0: &mut ()), (*(_1.0: &mut ()))]
DEBUG:<unknown>: places_conflict: Some(_1) vs. Some(_1)
DEBUG:<unknown>: place_element_conflict: DISJOINT-OR-EQ-LOCAL
DEBUG:<unknown>: places_conflict: Some((_1.0: &mut ())) vs. Some((_1.0: &mut ()))
DEBUG:<unknown>: place_element_conflict: DISJOINT-OR-EQ-FIELD
DEBUG:<unknown>: places_conflict: Some((*(_1.0: &mut ()))) vs. Some((*(_1.0: &mut ())))
DEBUG:<unknown>: place_element_conflict: DISJOINT-OR-EQ-DEREF
DEBUG:<unknown>: places_conflict: None vs. None
DEBUG:<unknown>: places_conflict: full borrow, CONFLICT
```

Reply:
I'm a bit confused by what you are showing me -- you are saying that if you *don't* use the "shallow" case, you get an error on the original test case?

Reply:
Yes. If I change it to `Deep` (or let it fall through to that case), then I get the error on the original test case. That's why I had it as `Shallow(None)` initially.